sdk/core/azure-core/src/uuid.cpp

```diff
@@ -18,84 +23,158 @@ static thread_local std::mt19937_64 randomGenerator(std::random_device{}());
 } // namespace
 #endif
 
+using Azure::Core::_internal::StringExtensions;
+
 namespace {
-static char ByteToHexChar(uint8_t byte)
+/*
+"00000000-0000-0000-0000-000000000000"
+         ^    ^    ^    ^
+ 000000000011111111112222222222333333
+ 012345678901234567890123456789012345
+ \______________ = 36 ______________/
+*/
+constexpr size_t UuidStringLength = 36;
+constexpr bool IsDashIndex(size_t i) { return i == 8 || i == 13 || i == 18 || i == 23; }
+
+constexpr std::uint8_t HexToNibble(char c) // does not check for errors
 {
-  if (byte <= 9)
+  if (c >= 'a')
   {
-    return '0' + byte;
+    return 10 + (c - 'a');
   }
 
-  AZURE_ASSERT_MSG(
-      byte >= 10 && byte <= 15,
-      "It is expected, for a valid Uuid, to have byte values, where each of the two nibbles fit "
-      "into a hexadecimal character");
+  if (c >= 'A')
+  {
+    return 10 + (c - 'A');
+  }
 
-  return 'a' + (byte - 10);
+  return c - '0';
+}
+
+constexpr char NibbleToHex(std::uint8_t nibble) // does not check for errors
+{
+  if (nibble <= 9)
+  {
+    return '0' + nibble;
+  }
+
+  return 'a' + (nibble - 10);
 }
 } // namespace
 
 namespace Azure { namespace Core {
   std::string Uuid::ToString() const
   {
-    std::string s(36, '-');
+    std::string s(UuidStringLength, '-');
 
-    for (size_t i = 0, j = 0; j < s.size() && i < UuidSize; i++)
+    for (size_t bi = 0, si = 0; bi < m_uuid.size(); ++bi)
     {
-      if (i == 4 || i == 6 || i == 8 || i == 10)
+      if (IsDashIndex(si))
       {
-        j++; // Add hyphens at the appropriate places
+        ++si;
       }
 
-      uint8_t highNibble = (m_uuid[i] >> 4) & 0x0F;
-      uint8_t lowNibble = m_uuid[i] & 0x0F;
-      s[j++] = ByteToHexChar(highNibble);
-      s[j++] = ByteToHexChar(lowNibble);
+      assert((si < UuidStringLength) && (si + 1 < UuidStringLength));
+
+      const std::uint8_t b = m_uuid[bi];
+      s[si] = NibbleToHex((b >> 4) & 0x0F);
+      s[si + 1] = NibbleToHex(b & 0x0F);
+      si += 2;
```

Comment:
We changed the few lines here, subtly.
I am curious, why?
For example:
Previously we incremented j++, now we don't and at the end do si += 2

Is that better?
And why remove the named locals for the nibbles, rather than pass the transformed value directly in the method call?

Trying to see the benefit. Imo, neither of those changes help readability and maintainability.

Reply:
It is pretty equal.
I didn't feel the need for there to be a local variable that is only used once, and I don't think it is hard to understand what it is used for, without making it a dedicated variable. It adds two lines (three if you add a space after), so function would take more lines on the screen increasing the mental resources you need to read it. Plus, previously the nobble variables were not `const`, so you as a reader might question if they get changed anywhere (does `ByteToHexChar()` take `uint8_t&` and modify them?), or if they are used in two places.

One `x+=2;` is potentially more effective than `x++; x++;` and even `++x; ++x;`, plus it makes an expression without side-effects (we're both assigning and incrementing on the same line; historically, this can sometimes bite you, so I avoid writing expressions with side effects just as a habit).

Reply:
When I initially saw the code, without named locals, it wasn't clear what the bitwise operation/expressions were doing.

Even though the locals were used only once, they described the semantic meaning. Now, a few months later, we'll have to re-page back and understand what these are, and which one is the high and low nibble.

I would recommend keeping the named locals (or at least add them as a comment next each bitwise expression line).

The compiler will do the right thing here, in any case, so expressiveness/readability for humans takes the win here.

Reply:
I have added more comments, and the follow-up PR I created is already merged, do you think I should open another for this? I would prefer, as you are back, if you made a PR and update it as you see fits the best, I think I would approve it (my only suggestion is to declare such variables `const`). It should be ok to update it later, as first this would likely be before the release, second it should be ok to update even after release.

Reply:
Yeah, sounds good.
I can address it once I am back, with some other changes.
Thanks!